pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -1121,6 +1133,72 @@ func (l *LibvirtDomainManager) getDomainSpec(dom cli.VirDomain) (*api.DomainSpec
 	return domainSpec, err
 }
 
+func (l *LibvirtDomainManager) GetDomainUpdateChan() chan DomainUpdateEvent {
+	return l.DomainUpdateChan
+}
+
+func (l *LibvirtDomainManager) completeMemoryDump() {
+	defer func() { <-l.memoryDumpInProgress }()
+	now := metav1.NewTime(time.Now())
+	memoryDumpInfo := &api.DomainMemoryDumpInfo{
+		DumpTimestamp: &now,
+	}
+	l.DomainUpdateChan <- DomainUpdateEvent{
+		MemoryDumpInfo: memoryDumpInfo,
+	}
+}
+
+func (l *LibvirtDomainManager) MemoryDump(vmi *v1.VirtualMachineInstance, dumpPath string) error {
+	select {
+	case l.memoryDumpInProgress <- struct{}{}:
+	default:
+		log.Log.Object(vmi).Infof("memory-dump is in progress")
+		return nil
+	}
+
+	if _, err := os.Stat(dumpPath); err == nil {
+		log.Log.Object(vmi).Infof("memory-dump for this command already completed")
+		return nil
+	}
+
+	go func() {
+		defer l.completeMemoryDump()
+
+		if err := l.memoryDump(vmi, dumpPath); err != nil {
+			log.Log.Object(vmi).Error(err.Error())
+		}
+	}()
+	return nil
+}
+
+func (l *LibvirtDomainManager) memoryDump(vmi *v1.VirtualMachineInstance, dumpPath string) error {
+	l.domainModifyLock.Lock()
+	defer l.domainModifyLock.Unlock()
```

Comment:
it seems like we needed this lock whenever we're performing a get+set on the domain spec. I'm uncertain if we want to hold this lock during the duration of the core dump. This lock would actually block the virt-handler's reconcile loop potentially if the same VM is synced again while the dump is taking place.

Reply:
If you think its safe to remove it I will

Reply:
the way you restructured this with the lock not being held during the dump looks good now